File: t/math_conformance.py
#!/usr/bin/env python3
"""Conformance and ownership checks for the shared Python colour maths."""

from __future__ import print_function

import importlib.util
import json
import math
import os
import sys


ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
BIN = os.path.join(ROOT, "usr", "bin")
if BIN not in sys.path:
    sys.path.insert(0, BIN)

import pgen_colour_math as math_core


def load_source(name):
    path = os.path.join(BIN, name + ".py")
    spec = importlib.util.spec_from_file_location(name + "_conformance", path)
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


with open(os.path.join(ROOT, "t", "fixtures", "math_conformance.json")) as stream:
    FIXTURE = json.load(stream)


checks = 0


def close(label, actual, expected, tolerance=2e-12):
    global checks
    checks += 1
    scale = max(1.0, abs(float(expected)))
    if abs(float(actual) - float(expected)) > tolerance * scale:
        raise AssertionError("{}: {!r} != {!r}".format(label, actual, expected))


def close_tree(label, actual, expected):
    if isinstance(expected, list):
        if len(actual) != len(expected):
            raise AssertionError("{} length differs".format(label))
        for index, value in enumerate(expected):
            close_tree("{}/{}".format(label, index), actual[index], value)
    else:
        close(label, actual, expected)


for row in FIXTURE["pq_encode"]:
    close("PQ encode {} nits".format(row["nits"]),
          math_core.pq_encode_nits(row["nits"], clamp_peak=True), row["signal"])
for row in FIXTURE["pq_decode"]:
    close("PQ decode {}".format(row["signal"]),
          math_core.pq_decode_nits(row["signal"]), row["nits"])

close_tree("matrix inverse",
           math_core.matrix3_inverse(FIXTURE["matrix"]), FIXTURE["inverse"])
close_tree("matrix vector",
           math_core.matrix3_vector_multiply(FIXTURE["matrix"], FIXTURE["vector"]),
           FIXTURE["product"])
checks += 1
if math_core.matrix3_inverse([[1, 2, 3], [4, 5, 6], [5, 7, 9]]) is not None:
    raise AssertionError("singular matrix (row3 = row1 + row2) must invert to None")
checks += 1
if math_core.bradford_adaptation([0.0, 0.0, 0.0],
                                 math_core.ICC_D50_WHITE) is not None:
    raise AssertionError("degenerate source white must yield None, not junk scaling")
# Zero is the one input the four runtimes do not agree on, so it cannot live
# in the shared fixture. Python evaluates the transfer function at zero rather
# than short-circuiting, as the C header does; Perl and the browser return
# exactly 0. Pinned here so the divergence is deliberate and visible.
close("PQ encode zero", math_core.pq_encode_nits(0.0),
      7.3095590257839665e-07)
close("PQ encode zero, peak clamped",
      math_core.pq_encode_nits(0.0, clamp_peak=True),
      7.3095590257839665e-07)
# Boundary policies: the icc_finetune combination must return exactly the peak
# past the denominator zero (signal ~1.99206), and the default bounded domain
# must clamp an over-unity signal to the 1.0 result rather than exploding
# through the floored denominator.
close("PQ decode past denominator zero, historical policy",
      math_core.pq_decode_nits(2.5, clamp_signal=False,
                               nonpositive_result=10000.0),
      10000.0, tolerance=0.0)
close("PQ decode over-unity signal, default clamp",
      math_core.pq_decode_nits(2.5), math_core.pq_decode_nits(1.0),
      tolerance=0.0)

close_tree("Bradford cone response",
           math_core.BRADFORD, FIXTURE["bradford_cone_response"])
close_tree("Bradford D65 to D50",
           math_core.bradford_adaptation(math_core.D65_WHITE,
                                         math_core.ICC_D50_WHITE),
           FIXTURE["bradford_d65_to_d50"])

# Independent Colour 0.4.7 reference values. The slightly wider bounds are
# deliberate: Main used seven-decimal RGB/XYZ coefficients and retaining them
# is required for serial parity. They remain much tighter than a colour-code
# threshold and make that compatibility decision visible.
for index, row in enumerate(FIXTURE["colour_0_4_7_ictcp_reference"]):
    actual = math_core.xyz_to_ictcp(row["xyz"])
    for component, expected in enumerate(row["ictcp"]):
        close("Colour ICtCp {}/{}".format(index, component),
              actual[component], expected, tolerance=2e-7)
for index, row in enumerate(FIXTURE["colour_0_4_7_delta_e_itp_reference"]):
    close("Colour Delta E ITP {}".format(index),
          math_core.delta_e_itp_xyz(row["first"], row["second"]),
          row["delta_e"], tolerance=2e-7)

builder = load_source("icc_profile_builder")
companion = load_source("icc_companion_lut")
finetune = load_source("icc_finetune")
repair = load_source("icc_b2a_repair")


def prior_finetune_de_itp(xyz_a, xyz_b):
    """The exact pre-consolidation expression, retained as a parity oracle."""
    xyz_to_rgb2020 = ((1.7166512, -0.3556708, -0.2533663),
                      (-0.6666844, 1.6164812, 0.0157685),
                      (0.0176399, -0.0427706, 0.9421031))
    rgb_to_lms = ((1688.0 / 4096, 2146.0 / 4096, 262.0 / 4096),
                  (683.0 / 4096, 2951.0 / 4096, 462.0 / 4096),
                  (99.0 / 4096, 309.0 / 4096, 3688.0 / 4096))

    def itp(xyz):
        rgb = [sum(xyz_to_rgb2020[row][column] * xyz[column]
                   for column in range(3)) for row in range(3)]
        lms = [sum(rgb_to_lms[row][column] * max(0.0, rgb[column])
                   for column in range(3)) for row in range(3)]
        encoded = [finetune.nits_to_pq(value) for value in lms]
        return (0.5 * encoded[0] + 0.5 * encoded[1],
                0.5 * (6610 * encoded[0] - 13613 * encoded[1]
                       + 7003 * encoded[2]) / 4096.0,
                (17933 * encoded[0] - 17390 * encoded[1]
                 - 543 * encoded[2]) / 4096.0)

    first, second = itp(xyz_a), itp(xyz_b)
    return 720.0 * math.sqrt(sum((a - b) ** 2
                                 for a, b in zip(first, second)))


state = 0xD17F1E
for index in range(512):
    values = []
    for _ in range(6):
        state = (1664525 * state + 1013904223) & 0xFFFFFFFF
        values.append(state / 4294967295.0 * 400.0)
    actual = finetune.de_itp(values[:3], values[3:])
    expected = prior_finetune_de_itp(values[:3], values[3:])
    checks += 1
    if actual != expected:
        raise AssertionError("fine-tune dE ITP differs at case {}: {!r} != {!r}"
                             .format(index, actual, expected))

ownership = (
    ("builder matrix multiply", builder.mat_mul, math_core.matrix3_multiply),
    ("builder matrix vector", builder.mat_vec_mul, math_core.matrix3_vector_multiply),
    ("builder PQ decode", builder.pq_to_nits, math_core.pq_decode_nits),
    ("builder PQ encode", builder.nits_to_pq, math_core.pq_encode_nits),
    ("builder table sampling", builder.sample_table, math_core.sample_uniform_table),
    ("builder smoothstep", builder.smoothstep, math_core.smoothstep),
    ("companion matrix multiply", companion.mat_mul, math_core.matrix3_multiply),
    ("companion matrix vector", companion.mat_vec, math_core.matrix3_vector_multiply),
    ("finetune matrix inverse", finetune.mat_inv, math_core.matrix3_inverse),
    ("finetune matrix multiply", finetune.mat_mul, math_core.matrix3_multiply),
    ("finetune matrix vector", finetune.mat_vec, math_core.matrix3_vector_multiply),
    ("finetune table sampling", finetune.sample_values,
     math_core.sample_uniform_table),
    ("finetune smoothstep", finetune.smoothstep, math_core.smoothstep),
    ("repair matrix vector", repair.mmul_b, math_core.matrix3_vector_multiply),
    ("repair table sampling", repair.sample_curve,
     math_core.sample_uniform_table),
)
for label, actual, expected in ownership:
    checks += 1
    if actual is not expected:
        raise AssertionError(label + " is not owned by pgen_colour_math")

print("{} Python colour-math conformance checks passed".format(checks))

File: t/math_conformance.js
'use strict';

const fs=require('fs');
const path=require('path');
const root=path.resolve(__dirname,'..');
const runtime=require(path.join(root,'usr/share/PGenerator/webui-colour-math.js'));
const fixture=JSON.parse(fs.readFileSync(path.join(__dirname,'fixtures/math_conformance.json'),'utf8'));
const measurementFixture=JSON.parse(fs.readFileSync(path.join(__dirname,'fixtures/measurement_conformance.json'),'utf8'));
const targetFixture=JSON.parse(fs.readFileSync(path.join(__dirname,'fixtures/calibration_target_conformance.json'),'utf8'));
const ciedeFixture=JSON.parse(fs.readFileSync(path.join(__dirname,'fixtures/ciede2000_sharma.json'),'utf8'));
const signalCodeFixture=JSON.parse(fs.readFileSync(path.join(__dirname,'fixtures/signal_code_conformance.json'),'utf8'));

let checks=0;
function close(label,actual,expected,tolerance){
 checks++;
 if(typeof actual!=='number'||typeof expected!=='number'||
    !Number.isFinite(actual)||!Number.isFinite(expected))
  throw new Error(label+': non-finite comparison '+actual+' != '+expected);
 const allowed=(tolerance==null?2e-12:tolerance)*Math.max(1,Math.abs(expected));
 if(Math.abs(actual-expected)>allowed) throw new Error(label+': '+actual+' != '+expected);
}
function exact(label,actual,expected){
 checks++;
 if(actual!==expected) throw new Error(label+': '+actual+' != '+expected);
}

for(const row of fixture.pq_encode){
 close('JS PQ encode '+row.nits,runtime.pqEncodeNormalized(row.nits),row.signal);
 close('JS chart PQ encode '+row.nits,runtime.meterChartPqEncodeNormalized(row.nits),row.signal);
}
for(const row of fixture.pq_decode)
 close('JS PQ decode '+row.signal,runtime.meterChartPqDecodeNormalized(row.signal),row.nits);

const hardZero=fixture.policy_rows.pq_encode_zero.find(row=>row.policy==='hard_zero');
exact('JS PQ encode zero',runtime.pqEncodeNormalized(0),hardZero.signal);
exact('JS chart PQ encode zero',runtime.meterChartPqEncodeNormalized(0),0);
exact('JS PQ encode negative',runtime.pqEncodeNormalized(-5),0);

for(const row of measurementFixture.cct_from_xy){
 checks++;
 const estimate=runtime.meterCctFromXy(row.x,row.y);
 const actual=estimate==null?0:Math.round(estimate);
 if(actual!==row.cct) throw new Error('JS CCT '+row.name+': '+actual+' != '+row.cct);
}
exact('JS CCT rejects non-finite chromaticity',runtime.meterCctFromXy(NaN,0.3),null);

const adapted=runtime.meterBradfordAdaptXyz(
 0.31,0.42,0.18,{x:0.3127,y:0.3290},{x:0.3457,y:0.3585});
for(const [index,key] of ['X','Y','Z'].entries())
 close('JS Bradford D65 to D50 '+key,adapted[key],
  [0.32546129634603371,0.42209381787880462,0.13879518502169955][index],2e-7);
const product=runtime.matrix3VectorMultiply(fixture.matrix,fixture.vector);
for(let index=0;index<3;index++)
 close('JS matrix-vector '+index,product[index],fixture.product[index]);

for(const [index,row] of fixture.colour_0_4_7_ictcp_reference.entries()){
 const actual=runtime.xyzToICtCp(...row.xyz);
 for(const [component,key] of ['I','T','P'].entries())
  close('JS ICtCp reference '+index+' '+key,actual[key],row.ictcp[component],2e-7);
}
for(const [index,row] of fixture.colour_0_4_7_delta_e_itp_reference.entries())
 close('JS Delta E ITP reference '+index,runtime.deltaEITP(...row.first,...row.second),row.delta_e,2e-7);

for(const [index,row] of ciedeFixture.pairs.entries()){
 const first={L:row.first[0],a:row.first[1],b:row.first[2]};
 const second={L:row.second[0],a:row.second[1],b:row.second[2]};
 close('JS CIEDE2000 Sharma pair '+index,runtime.deltaE2000(first,second),row.delta_e,5e-5);
}

const jndFirst={L:12.4,a:3.2,b:-4.1};
const jndSecond={L:10.9,a:2.7,b:-3.4};
close('JS JND standard-SL fallback',runtime.deltaE2000JND(jndFirst,jndSecond,0,0),1.2542310400237857);
close('JS JND positive luminance',runtime.deltaE2000JND(jndFirst,jndSecond,1,2),2.0730321454603882);
close('JS JND low-nit Barten scaling',runtime.deltaE2000JND(jndFirst,jndSecond,0.006,0.004),20.240937842592650);

for(const row of targetFixture){
 const context=runtime.calibrationTargetContext(row);
 if(!context) throw new Error('JS target context rejected '+row.name);
 checks++;
 if(!Object.isFrozen(context)) throw new Error('JS target context is mutable for '+row.name);
 const actual=row.caller_policy==='autocal_1d'
  ?runtime.targetLuminanceForContext(context,row.stimulus,row.white_y,row.black_y)
  :runtime.targetRelativeLuminanceForContext(context,row.signal,row.white_y,row.black_y);
 close('JS target context '+row.name,actual,Number(row.expected));
}

const dv=runtime.calibrationTargetContext({
 caller_policy:'autocal_1d',signal_mode:'dv',target_gamma:'st2084',sdr_signal_peak:100
});
exact('JS DV tunnel policy is explicit',dv.transfer_policy,'dv_gamma_2_2_tunnel');
exact('JS context schema is stable',dv.schema,'pgen-calibration-target-context-v1');
exact('JS context version is stable',dv.context_version,1);

for(const signal of [-0.1,0,0.0031308,0.5,1,1.1]){
 const encoded=runtime.srgbEncodeBounded(signal);
 const decoded=runtime.srgbDecodeBounded(encoded);
 close('JS bounded sRGB round trip '+signal,decoded,Math.max(0,Math.min(1,signal)));
}
close('JS unbounded sRGB preserves headroom',runtime.srgbDecodeUnbounded(runtime.srgbEncodeUnbounded(1.1)),1.1);

for(const row of signalCodeFixture.rows){
 const policy=runtime.signalCodePolicy({
  signal_mode:row.signal_mode,signal_range:row.signal_range,...(row.options||{})
 });
 checks++;
 if(!policy||!Object.isFrozen(policy)) throw new Error('JS signal policy rejected '+row.name);
 const actual=runtime.signalPercentToCode(policy,row.stimulus_percent);
 exact('JS signal-code fixture '+row.name,JSON.stringify(actual),JSON.stringify(row.expected));
 const fraction=runtime.codeToSignalFraction(policy,actual.code);
 checks++;
 if(!Number.isFinite(fraction)||fraction<0) throw new Error('JS signal fraction invalid '+row.name);
}

const appSource=fs.readFileSync(path.join(root,'usr/share/PGenerator/webui-app.js'),'utf8');
exact('browser preview constructs the shared signal policy',appSource.includes('const policy=signalCodePolicy(input);'),true);
exact('browser generic preview delegates shared conversion',appSource.includes('return meterPreviewSignalCode(percent,opts||{});'),true);
exact('browser SDR26 preview delegates shared conversion',appSource.includes('return meterPreviewSignalCode(slot,{lgAutocal26:true});'),true);
exact('browser bit depth never comes from white-code magnitude',appSource.includes('wc>0 && wc<=255'),false);

console.log(checks+' JavaScript colour-math conformance checks passed');
